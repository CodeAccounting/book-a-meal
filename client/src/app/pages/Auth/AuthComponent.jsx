import React, { Fragment, Component } from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import queryString from 'query-string';

import { urlPropTypes, authPropTypes } from '../../../helpers/proptypes';
import Header from '../../shared/Header';
import Footer from '../../shared/Footer';
import Form from '../../shared/Form';
import LinkBtn from '../../shared/Link';

/**
 * @exports
 * @class Auth
 * @extends Component
 * @classdesc Creates Auth Component
 * @returns {JSX} Auth Component
 */
class Auth extends Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    ...urlPropTypes,
    ...authPropTypes,
    changeUrl: PropTypes.func.isRequired
  }

  /**
   * @memberof Auth
   * @returns {nothin} nothing
   */
  componentWillMount() {
    const query = queryString.parse(this.props.location.search);
    const { role } = query;

    if (this.props.type === 'signup' && role === 'caterer') {
      return this.setState({ type: 'catererSignup' });
    }

    if (this.props.type === 'signup' && (!role || role === 'customer')) {
      return this.setState({ type: 'customerSignup' });
    }

    return this.setState({ type: 'signin' });
  }

  /**
   * @memberof Auth
   * @returns {object} Auth Meta Object
   */
  getMeta = () => {
    let btnText, para1, para2;

    switch (this.state.type) {
      case 'customerSignup':
        btnText = 'SIGN UP';
        para1 = <Fragment>Already have an account? Sign in <LinkBtn clickHandler={() => this.changeForm('signin')}>here</LinkBtn></Fragment>;
        para2 = <LinkBtn clickHandler={() => this.changeForm('catererSignup')}>Signup as a Caterer</LinkBtn>;
        break;
      case 'catererSignup':
        btnText = 'SIGN UP';
        para1 = <Fragment>Already have an account? Sign in <LinkBtn clickHandler={() => this.changeForm('signin')}>here</LinkBtn></Fragment>;
        para2 = <LinkBtn clickHandler={() => this.changeForm('customerSignup')}>Signup as a Customer</LinkBtn>;
        break;
      default:
        btnText = 'SIGN IN';
        para1 = <Fragment>{"Don't have an account, signup"} <LinkBtn clickHandler={() => this.changeForm('customerSignup')}>here</LinkBtn></Fragment>;
        para2 = <LinkBtn clickHandler={() => this.changeForm('catererSignup')}>Signup as a Caterer</LinkBtn>;
    }

    return {
      btnText,
      extra: (
        <Fragment>
          <p className="form-extra-info text-center">{para1}</p>
          <p className="text-center">Or</p>
          <p className="form-extra-info text-center">{para2}</p>
        </Fragment>)
    };
  }

  /**
   * @memberof Auth
   * @returns {string} Auth Form Title
   */
  getFormTitle = () => {
    switch (this.state.type) {
      case 'customerSignup':
        return 'Start Filling Your Belly';
      case 'catererSignup':
        return 'Start Serving Customers';
      default:
        return 'Welcome Back';
    }
  }

  /**
   * @memberof Auth
   * @param {string} type auth type
   * @desc changes active form
   * @returns {nothing} returns nothing
   */
  changeForm = (type) => {
    this.setState({ type });

    switch (type) {
      case 'customerSignup':
        return this.props.changeUrl('/signup?role=customer');
      case 'catererSignup':
        return this.props.changeUrl('/signup?role=caterer');
      default:
        return this.props.changeUrl('/signin');
    }
  }

  /**
   * @memberof Auth
   * @returns {JSX} Auth Component
   */
  render() {
    const meta = this.getMeta();
    const title = this.getFormTitle();
    const { from } = this.props.location.state ? this.props.location.state : { from: { pathname: '/' } };
    const newLocation = this.props.type === 'signin' ? from : '/';

    if (this.props.isAuthenticated) {
      return <Redirect to={newLocation} />;
    }

    return (
      <Fragment>
        <Header type="unauth" />
        <div className="main-wrapper auth-wrapper">
          <div className="auth">
            <div className="title">
              <h2>{title}</h2>
              <hr className="w-75" />
            </div>
            <Form {...this.props} type={this.state.type} meta={meta} />
          </div>
        </div>
        <Footer />
      </Fragment>
    );
  }
}

export default Auth;
