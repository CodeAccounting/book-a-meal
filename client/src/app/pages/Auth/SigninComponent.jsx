import React, { Fragment } from 'react';
import Header from '../../shared/Header';
import Footer from '../../shared/Footer';
import Form from '../../shared/Form';

const signinMeta = {
  btnText: 'SIGN IN',
  extra: (
    <div className="content">
      <p className="form-extra-info text-center">{"Don't have an account, signup"}
        <a href="./signup?role=user"> here</a>
      </p>
      <p className="text-center">Or</p>
      <p className="form-extra-info text-center">
        <a href="./signup?role=caterer">Signup as a Caterer</a>
      </p>
    </div>)
};

const Signin = props => (
  <Fragment>
    <Header />
    <div className="main-wrapper auth-wrapper">
      <div className="auth">
        <div className="title">
          <h2>Welcome Back</h2>
          <hr className="w-75" />
        </div>
        <Form {...props} type="signin" meta={signinMeta} />
      </div>
    </div>
    <Footer />
  </Fragment>
);

export default Signin;
