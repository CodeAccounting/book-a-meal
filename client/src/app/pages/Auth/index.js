import { connect } from 'react-redux';
import AuthComponent from '../Auth/AuthComponent';

const mapStateToProps = state => ({
  submitting: state.auth.loading,
  isAuthenticated: state.auth.isAuthenticated,
  id: state.auth.user.id,
  submitError: state.auth.error
});

export default connect(mapStateToProps)(AuthComponent);
