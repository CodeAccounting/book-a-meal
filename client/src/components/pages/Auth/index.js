import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import AuthComponent from '../Auth/AuthComponent';

const mapStateToProps = state => ({
  submitting: state.auth.loading,
  isAuthenticated: state.auth.isAuthenticated,
  id: state.auth.user.id,
  submitError: state.auth.error
});

const mapDispatchToProps = dispatch => ({
  changeUrl: link => dispatch(push(link))
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthComponent);
