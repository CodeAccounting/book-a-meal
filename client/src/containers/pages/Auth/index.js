import { connect } from 'react-redux';
import AuthComponent from '../../../components/pages/Auth';

const mapStateToProps = state => ({
  submitting: state.auth.loading,
  isAuthenticated: state.auth.isAuthenticated,
  id: state.auth.user.id,
  submitError: state.auth.error
});

export default connect(mapStateToProps)(AuthComponent);
