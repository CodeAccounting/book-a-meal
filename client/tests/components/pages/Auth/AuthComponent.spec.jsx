import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import AuthComponent from '../../../../src/components/pages/Auth/AuthComponent';
import Auth from '../../../../src/components/pages/Auth';
import { initialValues } from '../../../setup/mockData';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore(initialValues);

const unAuthStore = mockStore({
  ...initialValues,
  auth: { ...initialValues.auth, isAuthenticated: false }
});

const props = {
  submitting: false,
  isAuthenticated: false,
  submitError: null,
  changeUrl: jest.fn(),
};

const signinLocation = {
  pathname: '/signin',
  state: { from: { pathname: '/' } },
};

const catererSignupLocation = {
  pathname: '/signup',
  state: { from: { pathname: '/' } },
  search: '?role=caterer'
};

const customerSignupLocation = {
  pathname: '/signup',
  state: { from: { pathname: '/' } },
  search: '?role=customer'
};
const { now } = Date;

describe('AuthComponent', () => {
  beforeAll(() => {
    Date.now = jest.fn(() => 0);
  });

  afterAll(() => {
    Date.now = now;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders signin form correctly', () => {
    const wrapper = shallow(<AuthComponent {...props} location={signinLocation} type="signin" />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders customer\'s signup form correctly', () => {
    const wrapper = shallow(<AuthComponent {...props} location={customerSignupLocation} type="signup" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders caterer\'s signup form correctly', () => {
    const wrapper = shallow(<AuthComponent {...props} location={catererSignupLocation} type="signup" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders connected signin component correctly', () => {
    const comp = (<Provider store={unAuthStore}><Auth {...props} location={signinLocation} type="signin" /></Provider>);
    const mountedWrapper = mount(comp, rrcMock.get());

    expect(toJson(mountedWrapper)).toMatchSnapshot();
    mountedWrapper.unmount();
  });

  it('sets state on mount: signin', () => {
    const comp = (<AuthComponent {...props} location={customerSignupLocation} type="signin" />);
    const wrapper = shallow(comp);

    expect(wrapper.state().type).toEqual('signin');
  });

  it('sets state on mount: customerSignup', () => {
    const comp = (<AuthComponent {...props} location={customerSignupLocation} type="signup" />);
    const wrapper = shallow(comp);

    expect(wrapper.state().type).toEqual('customerSignup');
  });

  it('sets state on mount: catererSignup', () => {
    const comp = (<AuthComponent {...props} location={catererSignupLocation} type="signup" />);
    const wrapper = shallow(comp);

    expect(wrapper.state().type).toEqual('catererSignup');
  });

  it('sets state on mount: catererSignup', () => {
    const comp = (<Auth store={store} {...props} location={catererSignupLocation} type="catererSignup" />);
    const wrapper = mount(comp, rrcMock.get());

    expect(wrapper.find('Redirect')).toBeTruthy();
  });
});
