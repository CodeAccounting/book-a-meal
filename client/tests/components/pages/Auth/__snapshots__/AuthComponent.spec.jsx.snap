// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthComponent renders caterer's signup form correctly 1`] = `
<React.Fragment>
  <Header
    type="unauth"
  />
  <div
    className="main-wrapper auth-wrapper"
  >
    <div
      className="auth"
    >
      <div
        className="title"
      >
        <h2>
          Start Serving Customers
        </h2>
        <hr
          className="w-75"
        />
      </div>
      <Connect(Form)
        changeUrl={[MockFunction]}
        isAuthenticated={false}
        location={
          Object {
            "pathname": "/signup",
            "search": "?role=caterer",
            "state": Object {
              "from": Object {
                "pathname": "/",
              },
            },
          }
        }
        meta={
          Object {
            "btnText": "SIGN UP",
            "extra": <UNDEFINED>
              <p
                className="form-extra-info text-center"
              >
                <UNDEFINED>
                  Already have an account? Sign in 
                  <Link
                    className={null}
                    clickHandler={[Function]}
                    id={null}
                  >
                    here
                  </Link>
                </UNDEFINED>
              </p>
              <p
                className="text-center"
              >
                Or
              </p>
              <p
                className="form-extra-info text-center"
              >
                <Link
                  className={null}
                  clickHandler={[Function]}
                  id={null}
                >
                  Signup as a Customer
                </Link>
              </p>
            </UNDEFINED>,
          }
        }
        submitError={null}
        submitting={false}
        type="catererSignup"
      />
    </div>
  </div>
  <Footer />
</React.Fragment>
`;

exports[`AuthComponent renders connected signin component correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Auth)
    changeUrl={[MockFunction]}
    isAuthenticated={false}
    location={
      Object {
        "pathname": "/signin",
        "state": Object {
          "from": Object {
            "pathname": "/",
          },
        },
      }
    }
    submitError={null}
    submitting={false}
    type="signin"
  >
    <Auth
      changeUrl={[Function]}
      isAuthenticated={false}
      location={
        Object {
          "pathname": "/signin",
          "state": Object {
            "from": Object {
              "pathname": "/",
            },
          },
        }
      }
      submitError={null}
      submitting={false}
      type="signin"
    >
      <Header
        type="unauth"
      >
        <div
          className="header unauth"
        >
          <div
            className="navbar unauth-navbar"
          >
            <div
              className="page-title"
            >
              <h3>
                <a
                  href="/"
                >
                  BOOK-A-MEAL
                </a>
              </h3>
            </div>
          </div>
        </div>
      </Header>
      <div
        className="main-wrapper auth-wrapper"
      >
        <div
          className="auth"
        >
          <div
            className="title"
          >
            <h2>
              Welcome Back
            </h2>
            <hr
              className="w-75"
            />
          </div>
          <Connect(Form)
            changeUrl={[Function]}
            isAuthenticated={false}
            location={
              Object {
                "pathname": "/signin",
                "state": Object {
                  "from": Object {
                    "pathname": "/",
                  },
                },
              }
            }
            meta={
              Object {
                "btnText": "SIGN IN",
                "extra": <UNDEFINED>
                  <p
                    className="form-extra-info text-center"
                  >
                    <UNDEFINED>
                      Don't have an account, signup
                       
                      <Link
                        className={null}
                        clickHandler={[Function]}
                        id={null}
                      >
                        here
                      </Link>
                    </UNDEFINED>
                  </p>
                  <p
                    className="text-center"
                  >
                    Or
                  </p>
                  <p
                    className="form-extra-info text-center"
                  >
                    <Link
                      className={null}
                      clickHandler={[Function]}
                      id={null}
                    >
                      Signup as a Caterer
                    </Link>
                  </p>
                </UNDEFINED>,
              }
            }
            submitError={null}
            submitting={false}
            type="signin"
          >
            <Form
              changeUrl={[Function]}
              dispatch={[Function]}
              isAuthenticated={false}
              location={
                Object {
                  "pathname": "/signin",
                  "state": Object {
                    "from": Object {
                      "pathname": "/",
                    },
                  },
                }
              }
              meal={null}
              meta={
                Object {
                  "btnText": "SIGN IN",
                  "extra": <UNDEFINED>
                    <p
                      className="form-extra-info text-center"
                    >
                      <UNDEFINED>
                        Don't have an account, signup
                         
                        <Link
                          className={null}
                          clickHandler={[Function]}
                          id={null}
                        >
                          here
                        </Link>
                      </UNDEFINED>
                    </p>
                    <p
                      className="text-center"
                    >
                      Or
                    </p>
                    <p
                      className="form-extra-info text-center"
                    >
                      <Link
                        className={null}
                        clickHandler={[Function]}
                        id={null}
                      >
                        Signup as a Caterer
                      </Link>
                    </p>
                  </UNDEFINED>,
                }
              }
              submitError={null}
              submitting={false}
              type="signin"
            >
              <div>
                <form
                  onSubmit={[Function]}
                >
                  <SigninForm
                    handlers={
                      Object {
                        "handleBlur": [Function],
                        "handleChange": [Function],
                        "handleFocus": [Function],
                        "handleSubmit": [Function],
                      }
                    }
                    state={
                      Object {
                        "asyncValidating": false,
                        "error": Object {
                          "email": null,
                          "password": null,
                        },
                        "formValid": false,
                        "pristine": true,
                        "touched": Object {
                          "email": false,
                          "password": false,
                        },
                        "type": "signin",
                        "values": Object {
                          "email": "",
                          "password": "",
                        },
                      }
                    }
                    type="signin"
                  >
                    <RenderInput
                      handleBlur={[Function]}
                      handleChange={[Function]}
                      handleFocus={[Function]}
                      id="email"
                      label="Email Address"
                      meta={
                        Object {
                          "asyncValidating": false,
                          "error": null,
                          "touched": false,
                        }
                      }
                      name="email"
                      placeholder=""
                      required={true}
                      type="email"
                      value=""
                    >
                      <div
                        className="form-input"
                      >
                        <label
                          htmlFor="email"
                        >
                          Email Address
                          <span
                            className="danger"
                          >
                            *
                          </span>
                        </label>
                        <input
                          className=""
                          id="email"
                          name="email"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          placeholder=""
                          type="email"
                          value=""
                        />
                      </div>
                    </RenderInput>
                    <RenderInput
                      handleBlur={[Function]}
                      handleChange={[Function]}
                      handleFocus={[Function]}
                      id="password"
                      label="Password"
                      meta={
                        Object {
                          "error": null,
                          "touched": false,
                        }
                      }
                      name="password"
                      placeholder=""
                      required={true}
                      type="password"
                      value=""
                    >
                      <div
                        className="form-input"
                      >
                        <label
                          htmlFor="password"
                        >
                          Password
                          <span
                            className="danger"
                          >
                            *
                          </span>
                        </label>
                        <input
                          className=""
                          id="password"
                          name="password"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          placeholder=""
                          type="password"
                          value=""
                        />
                      </div>
                    </RenderInput>
                  </SigninForm>
                  <button
                    className="btn btn-pri btn-block"
                    disabled={true}
                  >
                    SIGN IN
                  </button>
                </form>
                <p
                  className="form-extra-info text-center"
                >
                  Don't have an account, signup
                   
                  <Link
                    className={null}
                    clickHandler={[Function]}
                    id={null}
                  >
                    <button
                      className="link-btn null"
                      id={null}
                      onClick={[Function]}
                    >
                      here
                    </button>
                  </Link>
                </p>
                <p
                  className="text-center"
                >
                  Or
                </p>
                <p
                  className="form-extra-info text-center"
                >
                  <Link
                    className={null}
                    clickHandler={[Function]}
                    id={null}
                  >
                    <button
                      className="link-btn null"
                      id={null}
                      onClick={[Function]}
                    >
                      Signup as a Caterer
                    </button>
                  </Link>
                </p>
              </div>
            </Form>
          </Connect(Form)>
        </div>
      </div>
      <Footer>
        <div
          className="footer"
        >
          <p>
            © 2018. Iveren Shaguy. All images were gotten from Google and are not mine.
          </p>
        </div>
      </Footer>
    </Auth>
  </Connect(Auth)>
</Provider>
`;

exports[`AuthComponent renders customer's signup form correctly 1`] = `
<React.Fragment>
  <Header
    type="unauth"
  />
  <div
    className="main-wrapper auth-wrapper"
  >
    <div
      className="auth"
    >
      <div
        className="title"
      >
        <h2>
          Start Filling Your Belly
        </h2>
        <hr
          className="w-75"
        />
      </div>
      <Connect(Form)
        changeUrl={[MockFunction]}
        isAuthenticated={false}
        location={
          Object {
            "pathname": "/signup",
            "search": "?role=customer",
            "state": Object {
              "from": Object {
                "pathname": "/",
              },
            },
          }
        }
        meta={
          Object {
            "btnText": "SIGN UP",
            "extra": <UNDEFINED>
              <p
                className="form-extra-info text-center"
              >
                <UNDEFINED>
                  Already have an account? Sign in 
                  <Link
                    className={null}
                    clickHandler={[Function]}
                    id={null}
                  >
                    here
                  </Link>
                </UNDEFINED>
              </p>
              <p
                className="text-center"
              >
                Or
              </p>
              <p
                className="form-extra-info text-center"
              >
                <Link
                  className={null}
                  clickHandler={[Function]}
                  id={null}
                >
                  Signup as a Caterer
                </Link>
              </p>
            </UNDEFINED>,
          }
        }
        submitError={null}
        submitting={false}
        type="customerSignup"
      />
    </div>
  </div>
  <Footer />
</React.Fragment>
`;

exports[`AuthComponent renders signin form correctly 1`] = `
<React.Fragment>
  <Header
    type="unauth"
  />
  <div
    className="main-wrapper auth-wrapper"
  >
    <div
      className="auth"
    >
      <div
        className="title"
      >
        <h2>
          Welcome Back
        </h2>
        <hr
          className="w-75"
        />
      </div>
      <Connect(Form)
        changeUrl={[MockFunction]}
        isAuthenticated={false}
        location={
          Object {
            "pathname": "/signin",
            "state": Object {
              "from": Object {
                "pathname": "/",
              },
            },
          }
        }
        meta={
          Object {
            "btnText": "SIGN IN",
            "extra": <UNDEFINED>
              <p
                className="form-extra-info text-center"
              >
                <UNDEFINED>
                  Don't have an account, signup
                   
                  <Link
                    className={null}
                    clickHandler={[Function]}
                    id={null}
                  >
                    here
                  </Link>
                </UNDEFINED>
              </p>
              <p
                className="text-center"
              >
                Or
              </p>
              <p
                className="form-extra-info text-center"
              >
                <Link
                  className={null}
                  clickHandler={[Function]}
                  id={null}
                >
                  Signup as a Caterer
                </Link>
              </p>
            </UNDEFINED>,
          }
        }
        submitError={null}
        submitting={false}
        type="signin"
      />
    </div>
  </div>
  <Footer />
</React.Fragment>
`;
